Mathlib/Computability/TuringDegrees.lean

```diff
@@ -0,0 +1,267 @@
+/-
+Copyright (c) 2024 Tanner Duve. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Tanner Duve, Elan Roth
+-/
+import Mathlib.Computability.Primrec
+import Mathlib.Computability.Partrec
+import Mathlib.Data.Part
+import Mathlib.Order.Antisymmetrization
+
+/-!
+# Oracle Computability and Turing Degrees
+
+This file defines a model of oracle computability, introduces Turing reducibility and equivalence,
+proves that Turing equivalence is an equivalence relation, and defines Turing degrees as the
+quotient under this relation.
+
+## Main Definitions
+
+- `RecursiveIn g f`:
+  An inductive definition representing that a partial function `f` is recursive in oracle `g`.
+- `TuringEquivalent`: A relation defining Turing equivalence between partial functions.
+- `TuringDegree`:
+  The type of Turing degrees, defined as equivalence classes under `TuringEquivalent`.
+
+## Notation
+
+- `f ≡ᵀ g` : `f` is Turing equivalent to `g`.
+
+## Implementation Notes
+
+The type of partial functions recursive in an oracle `g` is the smallest type containing
+the constant zero, the successor, projections, the oracle `g`, and is closed under
+pairing, composition, primitive recursion, and μ-recursion.
+
+## References
+
+* [Carneiro2018] Carneiro, Mario.
+  *Formalizing Computability Theory via Partial Recursive Functions*.
+  arXiv preprint arXiv:1810.08380, 2018.
+* [Odifreddi1989] Odifreddi, Piergiorgio.
+  *Classical Recursion Theory: The Theory of Functions and Sets of Natural Numbers,
+  Vol. I*. Springer-Verlag, 1989.
+* [Soare1987] Soare, Robert I. *Recursively Enumerable Sets and Degrees*. Springer-Verlag, 1987.
+* [Gu2015] Gu, Yi-Zhi. *Turing Degrees*. Institute for Advanced Study, 2015.
+
+## Tags
+
+Computability, Oracle, Turing Degrees, Reducibility, Equivalence Relation
+-/
+
+open Primrec Nat.Partrec Part
+
+/--
+The type of partial functions `f` that are recursive in an oracle `g` is the smallest type
+containing the constant zero, the successor, left and right projections, the oracle `g`,
+and is closed under pairing, composition, primitive recursion, and μ-recursion.
+
+Equivalently one can say that `f` is turing reducible to `g` when `f` is recursive in `g`.
+-/
+
+inductive RecursiveIn : (ℕ →. ℕ) → (ℕ →. ℕ) → Prop
+  | zero (g) : RecursiveIn (fun _ => 0) g
+  | succ (g) : RecursiveIn Nat.succ g
+  | left (g) : RecursiveIn (fun n => (Nat.unpair n).1) g
+  | right (g) : RecursiveIn (fun n => (Nat.unpair n).2) g
+  | oracle (g) : RecursiveIn g g
```

Comment:
I wonder if it makes sense to have multiple oracles here? That is, to change this to
```
inductive RecursiveIn (ι : Sort*) : (ℕ →. ℕ) → (oracles : ι → ℕ →. ℕ) → Prop
  | oracle (i) : RecursiveIn (oracles i) oracles
  -- etc
```
What's nice about this definition is that we can redefine `PartRec = RecursiveIn Empty`.

@digama0 might have more thoughts on whether this is actually a good idea.

Reply:
Makes sense to me. You could also use a set of oracles instead of an indexed family. (I met with @tannerduve at one point to discuss this project and I think I may even have suggested as much at the time.)

Reply:
Do you want to allow the program itself to choose which oracle to access, if you have countably many?

Reply:
What benefits does this have over implementing this by combining the oracles to one function and applying a single-function `RecursiveIn`? `Partrec` could be redefined as either `RecursiveIn fun _ ↦ ⊥` or `RecursiveIn fun _ ↦ 0`.

Reply:
Agreed @digama0  that a `Set` of oracles makes more sense than my indexed family, especially since this lives in `Prop`.

I don't know enough about the domain to know what to do if there are countably many oracles, or whether instead requiring the caller to combine them into a single oracle is sensible.

If you prefer this latter approach, can you add to the docstring a comment suggesting how oracles can be combined in this way?

Reply:
Oracles can't be combined in general. For finite oracles you can multiplex them but already for countable oracles you run into computability restrictions. For uncountable oracles it's simply impossible to combine them into one.